Expose cliPrint functions for use in debugging

etracer65 opened the pull request:
This is something I use a lot so I figured it would be helpful for others. With this change by simply by adding `#include "cli/cli.h"` to a file (if it's not already included) you can then use the various `cliPrint` functions to output debugging information to the CLI.

Exposes:
```
cliPrintLineFeed
cliPrint
cliPrintLine
cliPrintf
cliPrintLinef
```
Output is suppressed when the CLI is not open.

May interfere with the autocomplete initialization when first entering the CLI if your code is outputting data when the CLI first opens. But as this is only meant for debugging it shouldn't be much of a concern.

Using these calls in other modules **should only used for debugging** and not left in any finished code.

mikeller commented:
Not sure this is the right way to do this - I think to do this properly we should pull the console printing functionality into a compilation unit of its own, and then include this in CLI and a 'debug print' facility - perferably the debug print facility should also be done in a way that it has a 'master switch' define that will not build any of it if it is not defined, to make it easy to do clean release builds.

etracer65 commented:
Well this was meant to be a quick and dirty way of slapping some debugging in while testing. I wouldn't want to leave any of the debugging code in permanently which having some kind of master define implies. While certainly a more sophisticated debug/console print could be done that wasn't really the intent. Just trying to use what we already have since it works well and intuitive.

mikeller commented:
The problem is if we put in this facility, others will want to use it as well, and will probably be less careful about not leaving debug statements in (which will be hard to spot as they don't even say 'debug').
I am all for reuse, but I think a bit of refactoring would make this a lot better.

etracer65 commented:
Ok, that makes sense. How about separate `console_print` module that provides:
```
consolePrintLine
consolePrintLinef
...
```
Then `console_print.h` can include `cli.h` and the functions just call their related `cliPrint` functions. That way the `cliMode` check can be enforced in the `console_print.c` along with some master define to bound it. Then the user would only need to add `console_print.h` and make sure the master define was set. We could then possibly explicitly undefine the master define in the CI tests to catch cases where it got left (not sure if that's practical). Or maybe some other way to check for it in the CI tests.

But looking at this there might be one complication - some of the print functions are variant parameter based. So I don't know for example how we'd have `consolePrintf()` call `cliPrintf()` and pass the variant parameters. Might have to duplicate the function and just call the same backend `cliPrintfva()`

mikeller commented:
I'd probably rather use `cliDebugPrintLine` and similar to make it clear that these are meant for debugging.

We could probably use defines to do the renaming for us.

etracer65 commented:
@mikeller Any thoughts on how to flag/fail if the code is still included during CI? We basically need to test for `#ifdef USE_CLI_DEBUG_PRINT` and then trigger a `#error` and it probably makes sense to do that in the unit tests. I'm just not sure where we can globally test for it in the unit test structure. The include that remaps the function names is bounded so it will cause any use of the `cliDebugPrint...` functions to fail if `USE_CLI_DEBUG_PRINT` is not defined (which it won't be by default in the unit tests). But that will only get picked up if the unit test happens to compile the module that used the `cliDebugPrint...`.

The advantage of having it picked up by every unit test (if possible) is that the developer is more likely to get the error and realize they need to fix before submitting. Whereas if it was a specific "check for USE_CLI_DEBUG_PRINT" test they may not run into it unless they do a full `make test`. I'm just not sure it's possible as the unit tests don't share the same structure for the common places of setting up the define, like `target.h`, `common_pre.h`, `common_post.h`, etc.

I'm already generating an error if the new `cli_debug_print.h` is included but `USE_CLI_DEBUG_PRINT` is not defined as more crosschecking.

etracer65 commented:
I pushed an update to indicate where I'm heading. I would still like to figure out an automated way to flag/error in the unit tests if `USE_CLI_DEBUG_PRINT` is defined.

mikeller commented:
@etracer65: Not sure we need to have this check in _every_ unit test - having it picked up by one test will be good enough, as this (like many other 'sins' committed by developers) will be picked up in CI.

etracer65 commented:
So still trying to figure out a good way to make either tests or builds fail in CI if this code gets left in by the submitter. If a test includes the module where the debugging was added that will be handled and generate an error. But this won't catch cases where no unit test happens to utilize the affected module.

Not sure if this is possible, but could we have the CI build set a define for the actual target build as an indicator that this build is part of CI? Something like "CI_BUILD"? Then I could put in `common_post.h`:
```
#ifdef CI_BUILD
#undef USE_CLI_DEBUG_PRINT
#endif
```
and any modules that include the code will error.

mikeller commented:
@etracer65: You could probably add a check for this to `make/checks.mk` - they get run as part of the 'tests' Travis job.

etracer65 commented:
As an alternative, would it be possible to have the travis jobs that compile the targets run like:
```
make TARGET=<target_name> OPTIONS="CI_BUILD"
```
Then we could put in `common_post.h`:
```
#if defined(CI_BUILD) && defined(USE_CLI_DEBUG_PRINT)
#error "USE_CLI_DEBUG_PRINT must not be defined!"
#endif
```
Or move it out to a separate post check that can be used to check for and exclude code that we want to prevent in submitted builds. Having the code be aware at compile time that it's doing a CI build seems useful and makes it a little more transparent then hiding the check in a make file.

mikeller commented:
Probably an easier way to achieve this will be to add a warning that is shown whenever `USE_CLI_DEBUG_PRINT` is defined, like

```
#warning Do not use USE_CLI_DEBUG_PRINT for production builds.
```

This will give useful information **and** sort out CI builds.

ledvinap commented:
One possibility is to pass USE_CLI_DEBUG_PRINT (or better USE_DEBUG and protect other debug options in #ifdef) on command line ...

etracer65 commented:
@mikeller 

> Probably an easier way to achieve this will be to add a warning that is shown whenever `USE_CLI_DEBUG_PRINT` is defined, like
>
> ```
> #warning Do not use USE_CLI_DEBUG_PRINT for production builds.
> ```
>
>This will give useful information **and** sort out CI builds.

Well I thought this would be a good approach, but unfortunately putting the warning in someplace like `common_post.h` causes it to generate the warning literally hundreds of times. So it makes using this feature during development difficult as you're always distracted by the warnings and won't see any other real warnings that you'd care about.

@ledvinap I updated the PR initial comment to include the method of compiling using `OPTIONS="USE_CLI_DEBUG_PRINT"` as an alternative to adding the define to the code.

mikeller commented:
@etracer65:

> Well I thought this would be a good approach, but unfortunately putting the warning in someplace like common_post.h causes it to generate the warning literally hundreds of times.

Yeah, putting warnings into header files isn't a good approach. But sticking it into a compilation unit that is included in `COMMON_SRC` in `make/source.mk` does the trick nicely. We have been using `src/main/build/version.c` for this, but realistically we should probably move the checks to `src/main/build/build_config.c` (or else we should rip this file out as it is empty).

etracer65 commented:
Added the warning to `build/build_config.c` and seems to be a better solution only generating a single warning. So unobtrusive for development but will have the desired effect of failing the CI builds. Also moved the F1 and F3 unsupported warnings from `build/version.c` to `build/build_config.c` as suggested.
